File: lib/services/ozel_gostergeler_service.dart
import 'package:flutter/services.dart';
import 'package:csv/csv.dart';
import '../models/ozel_gosterge_data.dart';
import 'github_csv_service.dart';

class OzelGostergelerService {
  // Mevcut gösterge isimlerini al
  static Future<List<String>> getAvailableIndicators() async {
    try {
      final String csvData =
          await GitHubCSVService.loadCSVFromGitHub('ozel_gostergeler.csv');
      print('ozel_gostergeler.csv veri uzunluğu: ${csvData.length}');

      // Manuel olarak satırlara böl
      List<String> lines = csvData.split(RegExp(r'\r?\n'));
      print('ozel_gostergeler.csv toplam satır sayısı: ${lines.length}');

      if (lines.isEmpty) {
        throw Exception('ozel_gostergeler.csv dosyası boş');
      }

      // İlk satır header - gösterge isimleri 2. sütundan başlar
      String headerLine = lines[0];
      List<List<dynamic>> parsedHeader =
          const CsvToListConverter().convert(headerLine);
      List<dynamic> headerRow = parsedHeader.isNotEmpty ? parsedHeader[0] : [];

      List<String> indicators = [];
      // 2. sütundan başlayarak gösterge isimlerini al (index 1'den başlar)
      for (int i = 1; i < headerRow.length; i++) {
        String indicator = headerRow[i].toString().trim();
        if (indicator.isNotEmpty) {
          indicators.add(indicator);
        }
      }

      print('Bulunan göstergeler: $indicators');
      return indicators;
    } catch (e) {
      print('Göstergeler yükleme hatası: $e');
      throw Exception('Göstergeler yükleme hatası: $e');
    }
  }

  // Seçilen göstergenin günlük verilerini yükle
  static Future<OzelGostergeData> loadIndicatorData(
      String selectedIndicator) async {
    try {
      // Günlük veriler
      final dailyData = await _loadDailyData(selectedIndicator);

      // Aylık değişim verileri
      final monthlyData = await _loadMonthlyData(selectedIndicator);

      return OzelGostergeData(
        gostergeName: selectedIndicator,
        dailyValues: dailyData['values'],
        dates: dailyData['dates'],
        monthlyChanges: monthlyData['values'],
        monthlyDates: monthlyData['dates'],
      );
    } catch (e) {
      print('Gösterge verisi yükleme hatası: $e');
      throw Exception('Gösterge verisi yükleme hatası: $e');
    }
  }

  // Günlük verileri ozel_gostergeler.csv'den oku
  static Future<Map<String, dynamic>> _loadDailyData(
      String selectedIndicator) async {
    final String csvData =
        await GitHubCSVService.loadCSVFromGitHub('ozel_gostergeler.csv');

    // Manuel olarak satırlara böl
    List<String> lines = csvData.split(RegExp(r'\r?\n'));
    print('ozel_gostergeler.csv günlük veri satır sayısı: ${lines.length}');

    if (lines.isEmpty) {
      throw Exception('ozel_gostergeler.csv dosyası boş');
    }

    // Header satırından gösterge sütununu bul
    String headerLine = lines[0];
    List<List<dynamic>> parsedHeader =
        const CsvToListConverter().convert(headerLine);
    List<dynamic> headerRow = parsedHeader.isNotEmpty ? parsedHeader[0] : [];

    int indicatorColumnIndex = -1;
    for (int i = 0; i < headerRow.length; i++) {
      if (headerRow[i].toString().trim() == selectedIndicator) {
        indicatorColumnIndex = i;
        break;
      }
    }

    if (indicatorColumnIndex == -1) {
      throw Exception('Seçilen gösterge bulunamadı: $selectedIndicator');
    }

    print('Gösterge sütun indeksi: $indicatorColumnIndex');

    List<double> values = [];
    List<String> dates = [];

    // Veri satırlarını işle (1. sütun tarih, diğerleri veriler)
    for (int i = 1; i < lines.length; i++) {
      String line = lines[i].trim();
      if (line.isNotEmpty) {
        List<List<dynamic>> parsedLine =
            const CsvToListConverter().convert(line);
        if (parsedLine.isNotEmpty) {
          List<dynamic> row = parsedLine[0];

          if (row.length > indicatorColumnIndex) {
            // İlk sütun tarih
            String date = row[0].toString().trim();

            // Gösterge verisi
            String valueStr = row[indicatorColumnIndex].toString().trim();
            double value = double.tryParse(valueStr) ?? 0.0;

            if (date.isNotEmpty && value != 0.0) {
              dates.add(date);
              values.add(value);
            }
          }
        }
      }
    }

    print('Günlük veri sayısı: ${values.length}');
    return {
      'values': values,
      'dates': dates,
    };
  }

  // Aylık değişim verilerini ozelgostergeleraylik.csv'den oku
  static Future<Map<String, dynamic>> _loadMonthlyData(
      String selectedIndicator) async {
    final String csvData =
        await GitHubCSVService.loadCSVFromGitHub('ozelgostergeleraylik.csv');

    // Manuel olarak satırlara böl
    List<String> lines = csvData.split(RegExp(r'\r?\n'));
    print('ozelgostergeleraylik.csv satır sayısı: ${lines.length}');

    if (lines.isEmpty) {
      throw Exception('ozelgostergeleraylik.csv dosyası boş');
    }

    // Header satırından tarih sütunlarını al (3. sütundan başlar, index 2'den)
    String headerLine = lines[0];
    List<List<dynamic>> parsedHeader =
        const CsvToListConverter().convert(headerLine);
    List<dynamic> headerRow = parsedHeader.isNotEmpty ? parsedHeader[0] : [];

    List<String> monthlyDates = [];
    for (int i = 2; i < headerRow.length; i++) {
      String date = headerRow[i].toString().trim();
      if (date.isNotEmpty && date.contains('-')) {
        monthlyDates.add(date);
      }
    }

    // Seçilen göstergenin satırını bul (2. sütunda gösterge isimleri var)
    List<double> monthlyValues = [];

    for (int i = 1; i < lines.length; i++) {
      String line = lines[i].trim();
      if (line.isNotEmpty) {
        List<List<dynamic>> parsedLine =
            const CsvToListConverter().convert(line);
        if (parsedLine.isNotEmpty) {
          List<dynamic> row = parsedLine[0];

          if (row.length >= 2) {
            String rowIndicator = row[1].toString().trim();

            if (rowIndicator == selectedIndicator) {
              // Bu satırdan aylık değerleri al (3. sütundan başlayarak)
              for (int j = 2;
                  j < row.length && j < monthlyDates.length + 2;
                  j++) {
                double value = double.tryParse(row[j].toString()) ?? 0.0;
                monthlyValues.add(value);
              }
              break;
            }
          }
        }
      }
    }

    print('Aylık veri sayısı: ${monthlyValues.length}');
    return {
      'values': monthlyValues,
      'dates': monthlyDates,
    };
  }
}

File: src/services/OzelGostergelerService.ts
import Papa from 'papaparse';
import {GitHubCSVService} from './GitHubCSVService';
import {OzelGostergeData, OzelGostergeDataModel} from '../models/OzelGostergeData';

export class OzelGostergelerService {
  // Backwards-compatible aliases (page code expects these names)
  static async getIndicatorNames(): Promise<string[]> {
    return await this.getAvailableIndicators();
  }

  static async loadOzelGostergeData(selectedIndicator: string): Promise<OzelGostergeData> {
    return await this.loadIndicatorData(selectedIndicator);
  }

  static async getAvailableIndicators(): Promise<string[]> {
    try {
      const csvData = await GitHubCSVService.loadCSVFromGitHub(
        'ozel_gostergeler.csv'
      );

      const lines = csvData.split(/\r?\n/);
      if (lines.length === 0) {
        throw new Error('ozel_gostergeler.csv dosyası boş');
      }

      const parsedHeader = Papa.parse(lines[0], {
        header: false,
        skipEmptyLines: true,
      });
      const headerRow =
        parsedHeader.data.length > 0 && Array.isArray(parsedHeader.data[0])
          ? (parsedHeader.data[0] as any[])
          : [];

      const indicators: string[] = [];
      for (let i = 1; i < headerRow.length; i++) {
        const indicator = headerRow[i]?.toString().trim() || '';
        if (indicator) {
          indicators.push(indicator);
        }
      }

      return indicators;
    } catch (e) {
      throw new Error(`Göstergeler yükleme hatası: ${e}`);
    }
  }

  static async loadIndicatorData(
    selectedIndicator: string
  ): Promise<OzelGostergeData> {
    try {
      const dailyData = await this.loadDailyData(selectedIndicator);
      const monthlyData = await this.loadMonthlyData(selectedIndicator);

      // Return the model instance (page uses helper methods like getYearToDateChange).
      return new OzelGostergeDataModel(
        selectedIndicator,
        dailyData.values,
        dailyData.dates,
        monthlyData.values,
        monthlyData.dates
      );
    } catch (e) {
      throw new Error(`Gösterge verisi yükleme hatası: ${e}`);
    }
  }

  private static async loadDailyData(selectedIndicator: string): Promise<{
    values: number[];
    dates: string[];
  }> {
    const csvData = await GitHubCSVService.loadCSVFromGitHub(
      'ozel_gostergeler.csv'
    );

    const lines = csvData.split(/\r?\n/);
    if (lines.length === 0) {
      throw new Error('ozel_gostergeler.csv dosyası boş');
    }

    const parsedHeader = Papa.parse(lines[0], {
      header: false,
      skipEmptyLines: true,
    });
    const headerRow =
      parsedHeader.data.length > 0 && Array.isArray(parsedHeader.data[0])
        ? (parsedHeader.data[0] as any[])
        : [];

    let indicatorColumnIndex = -1;
    for (let i = 0; i < headerRow.length; i++) {
      if (headerRow[i]?.toString().trim() === selectedIndicator) {
        indicatorColumnIndex = i;
        break;
      }
    }

    if (indicatorColumnIndex === -1) {
      throw new Error(`Seçilen gösterge bulunamadı: ${selectedIndicator}`);
    }

    const values: number[] = [];
    const dates: string[] = [];

    for (let i = 1; i < lines.length; i++) {
      const line = lines[i].trim();
      if (line) {
        try {
          const parsed = Papa.parse(line, {
            header: false,
            skipEmptyLines: true,
          });
          if (
            parsed.data.length > 0 &&
            Array.isArray(parsed.data[0]) &&
            parsed.data[0].length > indicatorColumnIndex
          ) {
            const row = parsed.data[0] as any[];
            const date = row[0]?.toString().trim() || '';
            const valueStr = row[indicatorColumnIndex]?.toString().trim() || '';
            const value = parseFloat(valueStr) || 0.0;

            if (date && value !== 0.0) {
              dates.push(date);
              values.push(value);
            }
          }
        } catch (e) {
          // Skip parse errors
        }
      }
    }

    return {values, dates};
  }

  private static async loadMonthlyData(selectedIndicator: string): Promise<{
    values: number[];
    dates: string[];
  }> {
    const csvData = await GitHubCSVService.loadCSVFromGitHub(
      'ozelgostergeleraylik.csv'
    );

    const lines = csvData.split(/\r?\n/);
    if (lines.length === 0) {
      throw new Error('ozelgostergeleraylik.csv dosyası boş');
    }

    const parsedHeader = Papa.parse(lines[0], {
      header: false,
      skipEmptyLines: true,
    });
    const headerRow =
      parsedHeader.data.length > 0 && Array.isArray(parsedHeader.data[0])
        ? (parsedHeader.data[0] as any[])
        : [];

    const monthlyDates: string[] = [];
    for (let i = 2; i < headerRow.length; i++) {
      const date = headerRow[i]?.toString().trim() || '';
      if (date && date.includes('-')) {
        monthlyDates.push(date);
      }
    }

    const monthlyValues: number[] = [];

    for (let i = 1; i < lines.length; i++) {
      const line = lines[i].trim();
      if (line) {
        try {
          const parsed = Papa.parse(line, {
            header: false,
            skipEmptyLines: true,
          });
          if (
            parsed.data.length > 0 &&
            Array.isArray(parsed.data[0]) &&
            parsed.data[0].length >= 2
          ) {
            const row = parsed.data[0] as any[];
            const rowIndicator = row[1]?.toString().trim() || '';

            if (rowIndicator === selectedIndicator) {
              for (
                let j = 2;
                j < row.length && j < monthlyDates.length + 2;
                j++
              ) {
                const value = parseFloat(row[j]?.toString() || '0') || 0.0;
                monthlyValues.push(value);
              }
              break;
            }
          }
        } catch (e) {
          // Skip parse errors
        }
      }
    }

    return {values: monthlyValues, dates: monthlyDates};
  }

  /**
   * Yıllık CSV'den son değeri okur (yıllık değişim için)
   * @param indicatorName Gösterge adı
   * @returns Yıllık değişim değeri (son satırdaki değer)
   */
  static async getYearlyChange(indicatorName: string): Promise<number> {
    try {
      const csvData = await GitHubCSVService.loadCSVFromGitHub('özelgöstergeleryıllık.csv', true);
      const lines = csvData.split(/\r?\n/).filter(line => line.trim());
      
      if (lines.length < 2) {
        console.log('özelgöstergeleryıllık.csv dosyası yeterli veri içermiyor');
        return 0.0;
      }
      
      // İlk satır header
      const headerLine = lines[0];
      const parsedHeader = Papa.parse(headerLine, {
        header: false,
        skipEmptyLines: true,
      });
      
      if (parsedHeader.data.length === 0 || !Array.isArray(parsedHeader.data[0])) {
        console.log('özelgöstergeleryıllık.csv header parse edilemedi');
        return 0.0;
      }
      
      const headerRow = parsedHeader.data[0] as any[];
      
      // Gösterge adının sütun index'ini bul
      let columnIndex = -1;
      for (let i = 0; i < headerRow.length; i++) {
        if (headerRow[i]?.toString().trim() === indicatorName) {
          columnIndex = i;
          break;
        }
      }
      
      if (columnIndex === -1) {
        console.log(`özelgöstergeleryıllık.csv'de gösterge bulunamadı: ${indicatorName}`);
        return 0.0;
      }
      
      // Son satırı parse et (boş satırları atla)
      let lastLine = '';
      for (let i = lines.length - 1; i >= 0; i--) {
        const line = lines[i].trim();
        if (line && !line.startsWith(',')) {
          lastLine = line;
          break;
        }
      }
      
      if (!lastLine) {
        console.log('özelgöstergeleryıllık.csv dosyasında geçerli veri satırı bulunamadı');
        return 0.0;
      }
      
      const parsed = Papa.parse(lastLine, {
        header: false,
        skipEmptyLines: true,
      });
      
      if (parsed.data.length > 0 && Array.isArray(parsed.data[0])) {
        const row = parsed.data[0] as any[];
        if (row.length > columnIndex) {
          const valueStr = row[columnIndex]?.toString().trim() || '';
          if (valueStr) {
            const value = parseFloat(valueStr);
            if (!isNaN(value)) {
              return value;
            }
          }
        }
      }
      
      console.log(`özelgöstergeleryıllık.csv'den değer okunamadı: ${indicatorName}`);
      return 0.0;
    } catch (e) {
      console.log(`Yıllık değişim okuma hatası (${indicatorName}):`, e);
      return 0.0;
    }
  }
}

